add monitoring content (#337)
digital-asset/docs.daml.com: docs/2.7.0/docs/canton/usermanual/monitoring.rst

Monitoring
==========

Introduction
------------

> Observability (also known as “monitoring”) can determine if the Daml Enterprise solution is healthy or not healthy. If the state is not healthy, observability helps diagnose the root cause. There are three parts to observability: metrics, logs, and traces. These are described in this section.

Reviewer: ```suggestion
Observability (also known as “monitoring”) lets you determine if the Daml Enterprise solution is healthy or not. If the state is not healthy, observability helps diagnose the root cause. There are three parts to observability: metrics, logs, and traces. These are described in this section.
```

To avoid becoming overwhelmed by the number of metrics and log messages, follow these steps:

> - Read the shortcut to learning what is important, which is described below in the section :ref:`Hands-On with the Daml Enterprise - Observability Example <hands-on>`. It is a great starting point and an inspiration when building your metric monitoring.

Reviewer: ```suggestion
- Read the shortcut to learning what is important, which is described below in the section :ref:`Hands-On with the Daml Enterprise - Observability Example <hands-on>` as a starting point and inspiration when building your metric monitoring.
```

This differs from a situation where a single Daml transaction results in multiple actions at the same time, such as archiving and creating multiple contracts. In that case, a single trace encompasses the entire process, since it occurs as part of a single transaction rather than the result of an external process reacting to Daml events.

.. _traces:

Traces
~~~~~~

Traces contain operations that are each represented by a span. A trace is a directed acyclic graph (DAG) of spans, where the edges between spans are defined as parent/child relationships (the definitions come from the `Opentelemetry glossary <https://opentelemetry.io/docs/concepts/glossary/>`_).

Canton reports several types of traces. One example: every Canton console command that interacts with the Admin API starts a trace whose initial span last for the entire duration of the command, including the GRPC call to the specific Admin API endpoint.

.. figure:: ./images/ping-trace.jpg
   :width: 100%
   :alt: A graph showing the trace of a Canton ping containing 18 spans.

   Graph of a Canton ping trace containing 18 spans

> Traces of Daml command submissions are important. The trace illustrated in the figure results from performing a Canton ping using the console. It contains 18 spans. The ping is started by ``participant1``, and ``participant2`` is the target. The trace focuses on the message exchange through the sequencer without digging deep into the message handlers or further processing of transactions.

Author: ```suggestion
Traces of Daml command submissions are important. The trace illustrated in the figure results when you perform a Canton ping using the console. The ping is a smoke test that sends a Daml transaction (create Ping, exercise choice Pong, exercise choice Archive) to test a connection. It uses a particular smart contract that is preinstalled on every Canton participant. The command uses the Admin API to access a preinstalled application, which then issues Ledger API commands operating on this smart contract. In this example, the trace contains 18 spans. The ping is started by ``participant1``, and ``participant2`` is the target. The trace focuses on the message exchange through the sequencer without digging deep into the message handlers or further processing of transactions.
```

As part of phase 3, ``participant2`` receives a message (see span 8), and ``participant1`` also receives a message (see span 9). Both participants asynchronously validate the messages. ``participant2`` does not need to respond. Since it is only an observer, span 8 has no children. ``participant1`` responds, however, which is visible at span 10. There, it again makes a call to ``sequencer1``, which receives it at span 11.

At span 12, ``participant1`` receives a successful send response message that signals that its message to the mediator was successfully sequenced. This occurs as part of phase 4, where confirmation responses are sent to the mediator. The mediator receives it at span 13, and it validates the message (phase 5).

In spans 14 and 15, ``mediator1`` (now at phase 6) asks ``sequencer1`` to send the transaction result messages to the participants.

To end this round of the transaction protocol, ``participant1`` and ``participant2`` receive their messages at spans 16 and 17, respectively. The messages are asynchronously validated, and their projections of the virtual shared ledger are updated (phase 7).

> As mentioned, there are two other transaction submissions that are unlinked from this ping trace but are part of the operation. It is possible to find them.

Author: ```suggestion
As mentioned, there are two other transaction submissions that are unlinked from this ping trace but are part of the operation.
```

> The second one starts at a span titled ``admin-ping.processTransaction``, which is created by ``participant2``. The third one has the same name but is initiated by ``participant1``.

Author: Going to remove that particular sentence, since the next sentences mention their location.

.. _creating_dumps:

Health Dumps
------------

You should provide as much information as possible to receive efficient support. For this purpose, Canton implements an information-gathering facility that gathers key essential system information for support staff. If you encounter an error where you need assistance, please ensure the following:

- Start Canton in interactive mode, with the ``-v`` option to enable debug logging: ``./bin/canton -v -c <myconfig>``. This provides a console prompt.
- Reproduce the error by following the steps that previously caused the error. Write down these steps so they can be provided to support staff.
- After you observe the error, type ``health.dump()`` into the Canton console to generate a ZIP file.

This creates a dump file (``.zip``) that stores the following information:

- The configuration you are using, with all sensitive data stripped from it (no passwords).

> - An extract of the logfile. Overly sensitive data is not logged into log files.

Reviewer: ```suggestion
- An extract of the log file. Sensitive data is not logged into log files.
```

The precise way the state is computed is subject to change.

Here is an example monitoring configuration to place inside a node configuration object:

.. literalinclude:: /canton/includes/mirrored/enterprise/integration-testing/src/main/resources/include/health-monitoring.conf

.. note::

    The gRPC health server is configured per Canton node, not per process, as is the case for the HTTP health check server (see below). This means that the configuration must be inserted within a node's configuration object.

.. note::

    To support usage as a Kubernetes liveness probe, the health server exposes a service named ``liveness`` that should be targeted when `configuring a gRPC probe <https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-grpc-liveness-probe>`__.
    The latter service always returns ``SERVING``.

HTTP Health Check
~~~~~~~~~~~~~~~~~

> The ``canton`` process can optionally expose an HTTP endpoint indicating whether the process believes it is healthy. This may be used as an uptime check or as a `Kubernetes liveness probe <https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/>`__. If enabled, the ``/health`` endpoint will respond to a ``GET`` HTTP request with a 200 HTTP status code (if healthy) or 500 (if unhealthy, along with a plain text description of why it is unhealthy).

Reviewer: ```suggestion
Optionally, the ``canton`` process can expose an HTTP endpoint indicating whether the process believes it is healthy. This may be used as an uptime check or as a `Kubernetes liveness probe <https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/>`__. If enabled, the ``/health`` endpoint will respond to a ``GET`` HTTP request with a 200 HTTP status code (if healthy) or 500 (if unhealthy, along with a plain text description of why it is unhealthy).
```

Enabling the Prometheus Reporter
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> It is recommended to configure `Prometheus <https://prometheus.io>`__ for metrics reporting. Other reporters (jmx, graphite, and csv) are supported, but they are deprecated. Any such reporter should be migrated to Prometheus.

Reviewer: ```suggestion
`Prometheus <https://prometheus.io>`__ is recommended for metrics reporting. Other reporters (jmx, graphite, and csv) are supported, but they are deprecated. Any such reporter should be migrated to Prometheus.
```
